File: src/net/minecraft/client/render/entity/ProjectileEntityRenderer.cpp
#include "net/minecraft/client/render/entity/ProjectileEntityRenderer.hpp"
#include "net/minecraft/client/gl/GL11.hpp"
#include "net/minecraft/client/render/Tessellator.hpp"
#include "net/minecraft/client/render/entity/EntityRenderDispatcher.hpp"
#include "net/minecraft/entity/Entity.hpp"
namespace net::minecraft::client::render::entity {
ProjectileEntityRenderer::ProjectileEntityRenderer(int itemTextureIdIn) : itemTextureId(itemTextureIdIn) {}
void ProjectileEntityRenderer::render(const net::minecraft::Entity& entity, double x, double y, double z, float yaw,
                                      float tickDelta) {
  (void)entity;
  (void)yaw;
  (void)tickDelta;
  const gl::MatrixGuard matrix;
  gl::GL11::glTranslatef(static_cast<float>(x), static_cast<float>(y), static_cast<float>(z));
  gl::GL11::glScalef(0.5f, 0.5f, 0.5f);
  bindTexture("/gui/items.png");
  Tessellator& tessellator = Tessellator::INSTANCE;
  const float uMin = static_cast<float>((itemTextureId % 16 * 16) + 0) / 256.0f;
  const float uMax = static_cast<float>((itemTextureId % 16 * 16) + 16) / 256.0f;
  const float vMin = static_cast<float>((itemTextureId / 16 * 16) + 0) / 256.0f;
  const float vMax = static_cast<float>((itemTextureId / 16 * 16) + 16) / 256.0f;
  constexpr float size = 1.0f;
  constexpr float half = 0.5f;
  constexpr float quarter = 0.25f;
  if(dispatcher != nullptr) {
    gl::GL11::glRotatef(180.0f - dispatcher->yaw_, 0.0f, 1.0f, 0.0f);
    gl::GL11::glRotatef(-dispatcher->pitch_, 1.0f, 0.0f, 0.0f);
  }
  tessellator.startQuads();
  tessellator.normal(0.0f, 1.0f, 0.0f);
  tessellator.vertex(0.0f - half, 0.0f - quarter, 0.0, uMin, vMax);
  tessellator.vertex(size - half, 0.0f - quarter, 0.0, uMax, vMax);
  tessellator.vertex(size - half, size - quarter, 0.0, uMax, vMin);
  tessellator.vertex(0.0f - half, size - quarter, 0.0, uMin, vMin);
  tessellator.draw();
}
} // namespace net::minecraft::client::render::entity
#include "net/minecraft/client/entity/EntityClientRendererRegistration.hpp"
#include "net/minecraft/entity/projectile/thrown/EggEntity.hpp"
#include "net/minecraft/entity/projectile/thrown/SnowballEntity.hpp"
namespace net::minecraft::entity::projectile::thrown {
std::unique_ptr<::net::minecraft::client::render::entity::EntityRenderer> SnowballEntity::ClientRenderer::create() {
  return std::make_unique<::net::minecraft::client::render::entity::ProjectileEntityRenderer>(14);
}
std::unique_ptr<::net::minecraft::client::render::entity::EntityRenderer> EggEntity::ClientRenderer::create() {
  return std::make_unique<::net::minecraft::client::render::entity::ProjectileEntityRenderer>(12);
}
} // namespace net::minecraft::entity::projectile::thrown
namespace {
static ::net::minecraft::registry::RegisterEntityRenderer<::net::minecraft::entity::projectile::thrown::SnowballEntity>
    snowballRendererReg;
static ::net::minecraft::registry::RegisterEntityRenderer<::net::minecraft::entity::projectile::thrown::EggEntity>
    eggRendererReg;
} // namespace

File: mcp/src/net/minecraft/client/render/entity/ProjectileEntityRenderer.java
/*
 * Decompiled with CFR 0.1.1 (FabricMC 57d88659).
 * 
 * Could not load the following classes:
 *  net.fabricmc.api.EnvType
 *  net.fabricmc.api.Environment
 */
package net.minecraft.client.render.entity;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.render.Tessellator;
import net.minecraft.client.render.entity.EntityRenderer;
import net.minecraft.entity.Entity;
import org.lwjgl.opengl.GL11;

@Environment(value=EnvType.CLIENT)
public class ProjectileEntityRenderer
extends EntityRenderer {
    private int itemTextureId;

    public ProjectileEntityRenderer(int itemTextureId) {
        this.itemTextureId = itemTextureId;
    }

    public void render(Entity entity, double x, double y, double z, float yaw, float pitch) {
        GL11.glPushMatrix();
        GL11.glTranslatef((float)x, (float)y, (float)z);
        GL11.glEnable(32826);
        GL11.glScalef(0.5f, 0.5f, 0.5f);
        this.bindTexture("/gui/items.png");
        Tessellator tessellator = Tessellator.INSTANCE;
        float f = (float)(this.itemTextureId % 16 * 16 + 0) / 256.0f;
        float f2 = (float)(this.itemTextureId % 16 * 16 + 16) / 256.0f;
        float f3 = (float)(this.itemTextureId / 16 * 16 + 0) / 256.0f;
        float f4 = (float)(this.itemTextureId / 16 * 16 + 16) / 256.0f;
        float f5 = 1.0f;
        float f6 = 0.5f;
        float f7 = 0.25f;
        GL11.glRotatef(180.0f - this.dispatcher.yaw, 0.0f, 1.0f, 0.0f);
        GL11.glRotatef(-this.dispatcher.pitch, 1.0f, 0.0f, 0.0f);
        tessellator.startQuads();
        tessellator.normal(0.0f, 1.0f, 0.0f);
        tessellator.vertex(0.0f - f6, 0.0f - f7, 0.0, f, f4);
        tessellator.vertex(f5 - f6, 0.0f - f7, 0.0, f2, f4);
        tessellator.vertex(f5 - f6, 1.0f - f7, 0.0, f2, f3);
        tessellator.vertex(0.0f - f6, 1.0f - f7, 0.0, f, f3);
        tessellator.draw();
        GL11.glDisable(32826);
        GL11.glPopMatrix();
    }
}

